File: LC/Python/Hard/239.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        ans = []
        q = []
        window_start = 0
        
        for window_end, num in enumerate(nums):
            while q and nums[q[-1]] <= num:
                q.pop()
            q.append(window_end)
            
            if window_start > q[0]:
                q.pop(0)
            
            if window_end + 1 >= k:
                ans.append(nums[q[0]])
                window_start += 1
                
        return ans

File: LC/C++/Hard/239.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        std::vector<int> ans;
        std::deque<int> dq;
        int windowStart = 0;

        for (int windowEnd = 0; windowEnd < nums.size(); ++windowEnd) {
            while (!dq.empty() && nums[dq.back()] < nums[windowEnd]) {
                dq.pop_back();
            }
            dq.push_back(windowEnd);

            if (windowStart > dq[0]) {
                dq.pop_front();
            }

            if (windowEnd + 1 >= k) {
                ans.push_back(nums[dq[0]]);
                windowStart += 1;
            }
        }
        
        return ans;
    }
};